rasa/shared/core/domain.py

```diff
@@ -1858,6 +1885,34 @@ def __repr__(self) -> Text:
             f"{len(self.entities)} entities, {len(self.form_names)} forms"
         )
 
+    @staticmethod
+    def _collect_actions(actions: List[Any]) -> List[Text]:
+        result: List[Text] = []
+
+        for action in actions:
+            if isinstance(action, dict):
+                for action_name in action:
+                    result += [action_name]
+            elif isinstance(action, str):
+                result += [action]
+
+        return result
+
+    @staticmethod
+    def _collect_actions_which_explicitly_need_domain(actions: List[Any]) -> List[Text]:
+        result: List[Text] = []
+
+        for action in actions:
+            if isinstance(action, dict):
+                for action_name in action:
+                    action_config = action.get(action_name)
+                    if isinstance(action_config, dict):
+                        should_send_domain = action_config.get("send_domain")
+                        if should_send_domain:
+                            result += [action_name]
```

Comment:
I'd recommend this refactoring as the over-indentation can become difficult to read (plus the schema validation that we have in place would enforce that actions in the domain must be a list of strings or dicts):
```suggestion
                for action_name, action_config in action.items():
                    should_send_domain = action_config.get("send_domain", False)
                    if should_send_domain:
                        result += [action_name]
```